includes/reports/overview/data-provider/class-sensei-reports-overview-data-provider-courses.php

```diff
@@ -69,9 +76,14 @@ public function get_items( array $filters ): array {
 		}
 
 		$courses_query = new WP_Query( apply_filters( 'sensei_analysis_overview_filter_courses', $course_args ) );
+		remove_filter( 'posts_orderby', array( $this, 'add_orderby_custom_field_to_query' ), 10, 2 );
+
+		$course_args_no_pagination = $this->remove_pagination_arguments( $course_args );
+		$wp_all_courses            = new WP_Query( apply_filters( 'sensei_analysis_overview_filter_courses', $course_args_no_pagination ) );
+		$this->last_items_ids      = array_column( $wp_all_courses->posts, 'ID' );
+
```

Comment:
I would try to optimize this code a bit. Here are a couple of suggestions:

* You could query only the IDs when getting all the items.
* You could apply the query args filter only once.
* You could remove the `remove_pagination_arguments` method because it's used only once and it doesn't provide much benefit.

Here's a code example:

```php
$course_args   = apply_filters( 'sensei_analysis_overview_filter_courses', $course_args );
$courses_query = new WP_Query( $course_args );

remove_filter( 'posts_orderby', array( $this, 'add_orderby_custom_field_to_query' ), 10, 2 );

$all_courses_query = new WP_Query(
	array_merge(
		$course_args,
		[
			'posts_per_page' => -1,
			'fields'         => 'ids',
		]
	)
);

$this->all_item_ids = $all_courses_query->posts;
```

You could also consider extracting the logic to a `get_all_item_ids` method (which accepts the query args) if you want to keep the `get_items` method a bit cleaner.

The same could be applied to the other data provider classes.

Reply:
[here](https://github.com/Automattic/sensei/pull/5091/commits/fd884d4965e070bdd6d3d56a1aa96b35abb4b8c2)

Reply:
Sorry for the back and forward, but now thinking about it, I'm not sure if this class should be responsible for storing all the ids when calling the `get_items` method. It makes more sense to just call `get_items` again and get the ids when needed.

I think a better approach would be to move this logic to a parent class. One option would be to move it to `Sensei_Reports_Overview_List_Table_Abstract`. Here's an example patch:

```diff
diff --git a/includes/reports/overview/data-provider/class-sensei-reports-overview-data-provider-courses.php b/includes/reports/overview/data-provider/class-sensei-reports-overview-data-provider-courses.php
index c1c32b428..12a30eebb 100644
--- a/includes/reports/overview/data-provider/class-sensei-reports-overview-data-provider-courses.php
+++ b/includes/reports/overview/data-provider/class-sensei-reports-overview-data-provider-courses.php
@@ -59,6 +59,7 @@ class Sensei_Reports_Overview_Data_Provider_Courses implements Sensei_Reports_Ov
 			'post_status'      => array( 'publish', 'private' ),
 			'posts_per_page'   => $filters['number'],
 			'offset'           => $filters['offset'],
+			'fields'           => $filters['fields'] ?? '',
 			'orderby'          => $filters['orderby'] ?? '',
 			'order'            => $filters['order'] ?? 'ASC',
 			'suppress_filters' => 0,
diff --git a/includes/reports/overview/data-provider/class-sensei-reports-overview-data-provider-lessons.php b/includes/reports/overview/data-provider/class-sensei-reports-overview-data-provider-lessons.php
index 519ad2a21..cf172d1ca 100644
--- a/includes/reports/overview/data-provider/class-sensei-reports-overview-data-provider-lessons.php
+++ b/includes/reports/overview/data-provider/class-sensei-reports-overview-data-provider-lessons.php
@@ -63,6 +63,7 @@ class Sensei_Reports_Overview_Data_Provider_Lessons implements Sensei_Reports_Ov
 			'post_status'      => array( 'publish', 'private' ),
 			'posts_per_page'   => $filters['number'],
 			'offset'           => $filters['offset'],
+			'fields'           => $filters['fields'] ?? '',
 			'orderby'          => $filters['orderby'] ?? '',
 			'order'            => $filters['order'] ?? 'ASC',
 			'post__in'         => $course_lessons,
diff --git a/includes/reports/overview/list-table/class-sensei-reports-overview-list-table-abstract.php b/includes/reports/overview/list-table/class-sensei-reports-overview-list-table-abstract.php
index 635c94116..9fd28c5ea 100644
--- a/includes/reports/overview/list-table/class-sensei-reports-overview-list-table-abstract.php
+++ b/includes/reports/overview/list-table/class-sensei-reports-overview-list-table-abstract.php
@@ -37,6 +37,13 @@ abstract class Sensei_Reports_Overview_List_Table_Abstract extends Sensei_List_T
 	 */
 	protected $post_type = 'course';
 
+	/**
+	 * All the item ids.
+	 *
+	 * @var array
+	 */
+	protected $all_item_ids = [];
+
 	/**
 	 * Data provider for current report.
 	 *
@@ -71,12 +78,11 @@ abstract class Sensei_Reports_Overview_List_Table_Abstract extends Sensei_List_T
 	}
 
 	/**
-	 * Prepare the table with different parameters, pagination, columns and table elements
+	 * Get the filter arguments needed to get the items.
 	 *
-	 * @return void
-	 * @since  1.7.0
+	 * @return array
 	 */
-	public function prepare_items() {
+	private function get_filter_args(): array {
 		// Handle orderby.
 		// phpcs:ignore WordPress.Security.NonceVerification.Recommended -- No action, nonce is not required.
 		$orderby = sanitize_key( wp_unslash( $_GET['orderby'] ?? '' ) );
@@ -112,17 +118,43 @@ abstract class Sensei_Reports_Overview_List_Table_Abstract extends Sensei_List_T
 			$args['search'] = esc_html( $search );
 		}
 
-		$filters           = array_merge( $args, $this->get_additional_filters() );
-		$this->items       = $this->data_provider->get_items( $filters );
+		return array_merge( $args, $this->get_additional_filters() );
+	}
+
+	/**
+	 * Prepare the table with different parameters, pagination, columns and table elements
+	 *
+	 * @return void
+	 */
+	public function prepare_items() {
+		$filter_args       = $this->get_filter_args();
+		$this->items       = $this->data_provider->get_items( $filter_args );
 		$this->total_items = $this->data_provider->get_last_total_items();
 
 		$total_items = $this->total_items;
-		$total_pages = ceil( $total_items / $per_page );
+		$total_pages = ceil( $total_items / $filter_args['number'] );
 		$this->set_pagination_args(
 			array(
 				'total_items' => $total_items,
 				'total_pages' => $total_pages,
-				'per_page'    => $per_page,
+				'per_page'    => $filter_args['number'],
+			)
+		);
+	}
+
+	/**
+	 * Get all the item ids.
+	 *
+	 * @return array The post ids.
+	 */
+	protected function get_all_item_ids() {
+		return $this->data_provider->get_items(
+			array_merge(
+				$this->get_filter_args(),
+				[
+					'number' => -1,
+					'fields' => 'ids',
+				]
 			)
 		);
 	}
diff --git a/includes/reports/overview/list-table/class-sensei-reports-overview-list-table-courses.php b/includes/reports/overview/list-table/class-sensei-reports-overview-list-table-courses.php
index 628f9f45c..63d8fec6d 100644
--- a/includes/reports/overview/list-table/class-sensei-reports-overview-list-table-courses.php
+++ b/includes/reports/overview/list-table/class-sensei-reports-overview-list-table-courses.php
@@ -61,7 +61,7 @@ class Sensei_Reports_Overview_List_Table_Courses extends Sensei_Reports_Overview
 			return $this->columns;
 		}
 
-		$total_courses_ids       = $this->data_provider->get_all_items_ids();
+		$total_courses_ids       = $this->get_all_item_ids();
 		$this->total_courses_ids = $total_courses_ids;
 		$total_completions       = 0;
 		if ( ! empty( $total_courses_ids ) ) {

```

What do you think?

Reply:
Here is the initial change: https://github.com/Automattic/sensei/pull/5091/commits/046f74ba279ce41b0fd0868dc04cc6bff3c815ad

There are other commits (multiple ones) adding small changes to this commit and fixing the tests